src/main/java/com/puppycrawl/tools/checkstyle/checks/TrailingCommentCheck.java

```diff
@@ -200,76 +188,47 @@ public boolean isCommentNodesRequired() {
+        final String commentContent = ast.getFirstChild().getText();
+        final boolean isBeforeCommentLegal = format.matcher(beforeComment).find();
+
+        if (ast.getType() == TokenTypes.SINGLE_LINE_COMMENT) {
+            if (!isBeforeCommentLegal && !isCommentMatchLegalPattern(commentContent)) {
+                log(ast, MSG_KEY);
+            }
         }
-    }
-
-    @Override
-    public void beginTree(DetailAST rootAST) {
-        final Map<Integer, TextBlock> cppComments = getFileContents()
-                .getSingleLineComments();
-        final Map<Integer, List<TextBlock>> cComments = getFileContents()
-                .getBlockComments();
-        final Set<Integer> lines = new HashSet<>();
-        lines.addAll(cppComments.keySet());
-        lines.addAll(cComments.keySet());
-
-        for (Integer lineNo : lines) {
-            final String line = getLines()[lineNo - 1];
-            final String lineBefore;
-            final TextBlock comment;
-            if (cppComments.containsKey(lineNo)) {
-                comment = cppComments.get(lineNo);
-                lineBefore = line.substring(0, comment.getStartColNo());
+        else {
+            final DetailAST commentEnd = ast.getFirstChild().getNextSibling();
+            final String commentEndLine = getLine(commentEnd.getLineNo() - 1);
+            final int commentEndColumnNo = commentEnd.getColumnNo();
+            final String afterComment;
+            if (commentEndLine.charAt(commentEndColumnNo) == '*') {
+                afterComment = commentEndLine.substring(commentEndColumnNo + 2);
             }
             else {
-                final List<TextBlock> commentList = cComments.get(lineNo);
-                comment = commentList.get(commentList.size() - 1);
-                lineBefore = line.substring(0, comment.getStartColNo());
-
-                // do not check comment which doesn't end line
-                if (comment.getText().length == 1
-                        && !CommonUtil.isBlank(line
-                            .substring(comment.getEndColNo() + 1))) {
-                    continue;
-                }
+                afterComment = commentEndLine.substring(commentEndColumnNo + 2 + 1);
             }
-            if (!format.matcher(lineBefore).find()
-                && !isLegalComment(comment)) {
-                illegalLines.put(lineNo, comment.getStartColNo());
+            final boolean isCommentEndOfLine = CommonUtil.isBlank(afterComment);
+            if (isCommentEndOfLine) {
+                final boolean isMultiLine = commentEnd.getLineNo() > ast.getLineNo();
+                // Multiline comment cannot be legal
+                final boolean isCommentLegal = !isMultiLine
+                        && isCommentMatchLegalPattern(commentContent);
+                if (!isBeforeCommentLegal && !isCommentLegal) {
+                    log(ast, MSG_KEY);
+                }
```

Comment:
```
if (!isBeforeCommentLegal && !isCommentLegal) {
                    log(ast, MSG_KEY);
                }
```
This a bit duplicates logic on line 197, I suggest to extract this below your big if-else block.
To be more specific, do something like
```
boolean isMultiline;
if (ast.getType() == TokenTypes.SINGLE_LINE_COMMENT) isMultiline = false;
else {...}
if (!isBeforeCommentLegal && !isCommentMatchLegalPattern(commentContent) && isMultiline) {
                    log(ast, MSG_KEY);
                }
```

Reply:
I agree. Duplicate logic removed